Design/indexing_rollout_notes.md

```diff
@@ -0,0 +1,209 @@
+
+
+## General goals
+
+- Quite a few algorithms take an input workspace, extract information about spectrum numbers and detector IDs, and initialize a new workspace based on those data. Example include `SumSpectra`, `DiffractionFocussing`, and `ExtractSpectra`.
+  One goal of the indexing library is to remove explicit uses of spectrum numbers and detector IDs in as much code as possible.
+  Since we need to refactor algorithms similar to those mentioned above, it is probably a good idea to do it properly right away.
+  That is, we should *not* just refactor the manipulation of spectrum numbers and detector IDs via `IndexInfo`.
+  Instead, we should provide functionality for grouping and extraction of spectra to create a new `IndexInfo`, without the need for explicit handling of spectrum numbers and detector IDs.
+
+  As an example, consider
+
+  ```cpp
+  namespace Indexing {
+  IndexInfo extract(const IndexInfo &, const SpectrumIndexSet &);
+  IndexInfo group(const IndexInfo &, const std::vector<SpectrumIndexSet> &);
+  }
+  ```
+
+- Should we set default detector IDs?
+  This is was workspaces are doing currently, but it does not really make sense.
+  Probably we should find a better way if setting sensible detector IDs once an instrument has been set.
+  This is something that potentially needs to be decided together with other decisions regarding Instrument-2.0.
+
+
+## NearestNeighbours
+
+- Uses `specnum_t` as map key.
+- Should simply use spectrum index?
+- Would that make trouble with MPI? Need to have a detailed look.
+- See issue [#17372](https://github.com/mantidproject/mantid/issues/17372).
+
+
+## Logging and exception messages
+
+- Frequently spectrum numbers are included in log messages or exceptions.
+
+- Probably the easiest (overall) solution is to add means of getting this information automatically, e.g., with a helper function
+
+  ```cpp
+  namespace Logging {
+  std::string spectrumNumber(const MatrixWorkspace &workspace, size_t index);
+  }
+  ```
+
+## File- and UI output
+
+Similar to logging, there are also several algorithms that write, e.g., detector IDs to a file.
+Maybe a similar or even the same helper function as suggested above in the logging section could be used.
+
+
+## Obtaining detector information
+
+Frequently, detector IDs are used to obtain detector pointers or other detector-related information from the instrument.
+The IDs are not actually required here.
+Refactoring would be significantly easier (and would avoid the need to revisit and incrementally refactor) if such information was available via, e.g., `MatrixWorkspace::SpectrumInfo`.
+
+Commonly used functionality includes:
+
+```cpp
+class SpectrumInfo {
+  bool isMasked(const size_t index) const;
+  bool isMonitor(const size_t index) const;
+  double l2(const size_t index) const;
+  Detector_sptr uniqueDetector(const size_t index) const; // Throws if no or more than 1 detector
```

Comment:
Yes, I think there are 3 reasons:

- As long as `SpectrumInfo` is just a wrapper around Instrument-1.0 detectors do not exist permanently, but a parametrized detectors is allocated on demand. This can be buffered in `SpectrumInfo`. If we would return a reference there is some danger of invalidating it by requesting information for another index. That is, returning a `shared_ptr` makes this a bit safer. Later it could (and should) be changed to a reference.
- There might still be other client code that requires a `shared_ptr`, which we also need to clean up first.
- Copy and paste ;-)